### DISCUSSION TITLE
tutorials/3d/using_gridmaps

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#149] [User-contributed notes]
# tutorials/3d/using_gridmaps

Introduction: Gridmaps are a tool for creating 3D game levels, similar to the way TileMap works in 2D. You start with a predefined collection of 3D meshes (a MeshLibrary) that can be placed on a gr...

https://docs.godotengine.org/en/stable/tutorials/3d/using_gridmaps.html

**Comment by GreenCloversGames** [comment]
I feel this should include a section about baking lights into a grid map. I didn't realize that it was something you could do until I spent more time getting into the gridmap source code.

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot-docs/pull/11189

**Comment by elamaunt** [comment]
How can we add a custom logic to every cell in the gridmap? Animations, code.. etc. It is not clear

**Reply by Calinou** [reply]
Individual GridMap cells are not scenes, so they cannot have scripts attached to them. Instead, you can detect collisions on the GridMap, pass the collision position to [`GridMap.local_to_map()`](https://docs.godotengine.org/en/stable/classes/class_gridmap.html#class-gridmap-method-map-to-local), then use [`GridMap.get_cell_item()`](https://docs.godotengine.org/en/stable/classes/class_gridmap.html#class-gridmap-method-get-cell-item) to figure out which cell type was collided with. If the cell ID matches a certain cell type, run specific code using an `if` or `match` statement.

GridMap cells also can't make use of AnimationPlayer, but you can use a material with a [custom shader](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/spatial_shader.html) applied to the MeshLibrary item to display some kind of animation.

**Comment by petersgiles** [comment]
how do I fix the origins of individual meshes in my mesh library? for example if I want a wall mesh to transform to z-2m so it on the edge of the placement cube where do i set that?

**Comment by massenzio** [comment]
hi there, is possible during the creation of a map using gridmap to only see a selected level?
I mean, when i make a multilevel map using just one gridmap at the moment is really difficult to edit the part covered by upper levels...
the only way i found is to separate level in different gridmap but maybe there is a better way i didnt know?

**Reply by Calinou** [reply]
> hi there, is possible during the creation of a map using gridmap to only see a selected level?

This is only possible by using separate GridMap nodes and toggling their visibility, as you mentioned. If you need this on a single GridMap node, please open a [proposal](https://github.com/godotengine/godot-proposals) for this.
